File: src/chapters/matrix-factorization/content/code/lu.go
package main

import "fmt"

// Doolittle factorization A = L U (L unit-lower, U upper).
func luDoolittle(A [][]float64) ([][]float64, [][]float64) {
	n := len(A)
	L := make([][]float64, n)
	U := make([][]float64, n)
	for i := range L {
		L[i] = make([]float64, n)
		U[i] = make([]float64, n)
		L[i][i] = 1
	}
	for i := 0; i < n; i++ {
		for j := i; j < n; j++ {
			s := A[i][j]
			for k := 0; k < i; k++ {
				s -= L[i][k] * U[k][j]
			}
			U[i][j] = s
		}
		for j := i + 1; j < n; j++ {
			s := A[j][i]
			for k := 0; k < i; k++ {
				s -= L[j][k] * U[k][i]
			}
			L[j][i] = s / U[i][i]
		}
	}
	return L, U
}

func main() {
	A := [][]float64{{1, -2, -2, -2}, {2, -1, 2, 4}, {-1, 2, 3, -4}, {-2, 1, 4, -2}}
	L, U := luDoolittle(A)
	for _, row := range L {
		fmt.Println(row)
	}
	for _, row := range U {
		fmt.Println(row)
	}
}

File: src/chapters/matrix-factorization/content/code/lu.rs
// Doolittle factorization A = L U (L unit-lower, U upper).
fn lu_doolittle(a: &[Vec<f64>]) -> (Vec<Vec<f64>>, Vec<Vec<f64>>) {
    let n = a.len();
    let mut l = vec![vec![0.0; n]; n];
    let mut u = vec![vec![0.0; n]; n];
    for i in 0..n {
        l[i][i] = 1.0;
        for j in i..n {
            let mut s = a[i][j];
            for k in 0..i { s -= l[i][k] * u[k][j]; }
            u[i][j] = s;
        }
        for j in i + 1..n {
            let mut s = a[j][i];
            for k in 0..i { s -= l[j][k] * u[k][i]; }
            l[j][i] = s / u[i][i];
        }
    }
    (l, u)
}
fn main() {
    let a = vec![
        vec![1.0, -2.0, -2.0, -2.0],
        vec![2.0, -1.0, 2.0, 4.0],
        vec![-1.0, 2.0, 3.0, -4.0],
        vec![-2.0, 1.0, 4.0, -2.0],
    ];
    let (l, u) = lu_doolittle(&a);
    for row in &l { println!("{:?}", row); }
    for row in &u { println!("{:?}", row); }
}
